### PR TITLE
Henryh/add modality x space time masking

### PR DESCRIPTION
**Pull request by Hgherzog**
This pull request introduces several updates across the codebase with a focus on enhancing masking strategies, adding utility methods, and improving error handling. The most significant changes include the addition of a new `ModalityCrossMaskingStrategy` and its variants, utility functions for bandset indices, and improvements to masking logic. Below is a categorized summary of the changes:

### New Features

* **New Masking Strategies**: Introduced `ModalityCrossMaskingStrategy` and its variants (`ModalityCrossSpaceMaskingStrategy`, `ModalityCrossTimeMaskingStrategy`, `ModalityCrossSpaceTimeMaskingStrategy`) to support advanced masking techniques for encoding and decoding bandsets across modalities. These strategies include logic for filtering bandset indices, selecting encoded/decoded bandsets, and applying masks based on spatial and temporal conditions. (`[helios/train/masking.pyR711-R989](diffhunk://#diff-e094e660dfe63384592faeef8bfefd5617a64e6b41538613263a904c747d4620R711-R989)`)

* **Powerset Utility**: Added a utility function `powerset` to generate all non-empty combinations of bandsets for masking strategies. This is complemented by the `ALL_BANDSET_IDXS` and `ALL_BANDSET_POWSET` constants, which define and precompute bandset combinations. (`[helios/train/masking.pyR22-R39](diffhunk://#diff-e094e660dfe63384592faeef8bfefd5617a64e6b41538613263a904c747d4620R22-R39)`)

### Utility Enhancements

* **Channel Index Retrieval**: Added `get_channel_indices_for_bandset` method in `helios/data/constants.py` to fetch channel indices for a specific bandset, improving code modularity and reusability. (`[helios/data/constants.pyR134-R137](diffhunk://#diff-f029cf9aa24d9fc94f848352d93763a779b845dbf5e3e2e1044f93a63d0b5d26R134-R137)`)

### Bug Fixes and Improvements

* **Masking Logic Update**: Replaced `.contiguous()` with `.clone()` in `apply_mask` to ensure the integrity of the mask tensor during reshaping operations. (`[[1]](diffhunk://#diff-e094e660dfe63384592faeef8bfefd5617a64e6b41538613263a904c747d4620L391-R410)`, `[[2]](diffhunk://#diff-e094e660dfe63384592faeef8bfefd5617a64e6b41538613263a904c747d4620L510-R529)`)

* **Error Handling**: Added a warning message in `compute` to handle cases where no decoder tokens are present in a sample, improving debugging capabilities. (`[helios/train/loss.pyL164-R170](diffhunk://#diff-b14a53a97abceeb9bb0cc22b253e5341322adb96fc948a9eb1b839f46b362f44L164-R170)`)

### CODE REVIEW

File: helios/train/loss.py [hunk 1]
```diff
@@ -161,11 +161,22 @@ def compute(
         target = F.normalize(target, p=2, dim=-1)
 
         count = (all_masks == MaskValue.DECODER.value).sum(dim=-1)
-
+        # log if any of the count values are 0
+        if any(count == 0):
+            logger.warning("Some samples have no decoder tokens.")
+        if all(count == 1):
```
Comment (Hgherzog) [inline helios/train/loss.py]: I think it is better to just exclude latlon

File: helios/data/constants.py [hunk 2]
```diff
@@ -131,6 +131,10 @@ def bandsets_as_indices(self) -> list[list[int]]:
             offset += num_bands
         return indices
 
+    def get_channel_indices_for_bandset(self, bandset_idx: int) -> list[int]:
```
Comment (pjreddie) [inline helios/data/constants.py]: is this ever used?

File: helios/train/masking.py [hunk 3]
```diff
@@ -689,6 +708,285 @@ def apply_mask(
...
+            masked_modality_name = MaskedHeliosSample.get_masked_modality_name(modality)
+            modality_spec = Modality.get(modality)
+            modality_num_bandsets = modality_spec.num_band_sets
+            modality_mask = masked_batch_dict[masked_modality_name]
+            # For static in space data ignore all previous masking decisions and clam
+            if self.overide_random_mask_condition(modality_spec):
+                logger.info(
+                    f"Clamping {modality} to min {MaskValue.TARGET_ENCODER_ONLY.value} for all bandsets"
+                )
+                modality_mask = torch.clamp(
```
Comment (pjreddie) [inline helios/train/masking.py]: I feel like it's a little weird to rely on the ordering of masking values for this, I'm not sure the order should have semantics

Comment (pjreddie) [reply helios/train/masking.py]: maybe something like torch.where(mask==ONLINE_ENCODER, TARGET_ENCODER, mask)

but do we want to be setting them to TARGET_ENCODER? why not set them to DECODER?

File: helios/data/constants.py (hunk 2 above)
Comment (Hgherzog) [reply helios/data/constants.py]: I realized I don't need this will remove it.

File: helios/train/masking.py (hunk 3 above)
Comment (Hgherzog) [reply helios/train/masking.py]: Oh totally, great catch. This also could interfere with missing masks for non spatial or temporal modalities. let me rework this

Comment (Hgherzog) [reply helios/train/masking.py]: @pjreddie  lmk what you think of the new implementation. Also, the reason we want to set the not encode and not decoded to target_encoder is because we want to preserve whatever happened in the original mask (ie for space or time)

File: tests/unit/train/test_masking.py [hunk 4]
```diff
@@ -801,3 +802,109 @@ def test_random_range_masking() -> None:
...
+        max_encoding_bandsets=3,
+        encode_ratio=0.1,
+        decode_ratio=0.75,
+    )
+    masked_sample = strategy.apply_mask(batch, patch_size=patch_size)
+    logger.info(f"masked_sample: {masked_sample}")
+    # Check that the worldcover mask has the expected values
+    # Check that latlon mask has the expected values
+    expected_latlon_mask = torch.tensor([[0], [0], [0], [0]])
+    expected_worldcover_mask = torch.tensor(
```
Comment (pjreddie) [inline tests/unit/train/test_masking.py:840]: how do you know? is it just seeded?

Comment (Hgherzog) [reply tests/unit/train/test_masking.py:840]: yea conftest seeds all the randoms
